### PR TITLE
Es6/tests and examples 2

### PR DESCRIPTION
**Pull request by wwwtyro**

### CODE REVIEW

**Comment by wwwtyro** [comment]
Hey @ibgreen, can you try lesson 1 with this branch and let me know how this behavior is diverging from your intent? I'm trying to puzzle out the intended recursiveLoad functionality:

* Should we be returning a Promise from recursiveLoad? If so, is there some way to handle the recursive await call inside recursiveLoad and still return a Promise? I tried my hand at it, but wasn't able to figure out how, yet.
* Should we be trying to handle the different kinds of `#include`s that might have, for example, a fromShaderIds-type reference? Right now it appears to assume an XHR request is required.

**Comment by ibgreen** [comment]
Hi Rye, 

I don't care too much about the recursive load function, that code was too complex and apps can now use the "glslify" transform instead. So, if the examples don't require recursiveLoad, it's ok to comment it out for now, and/or modify the examples.

That said, the async functions all return promises per definition, and loading functions should too. 

Ib

**Comment by wwwtyro** [comment]
Oh, awesome, wasn't aware we were moving to glslify already. Great news! :+1:

I should be able to continue from here, then. Thanks Ib!

**Comment by wwwtyro** [comment]
Hi @ibgreen, lessons through 16 are working; should I continue with the examples?

**Comment by vicapow** [comment]
I think @ibgreen might be OOO. Maybe @philogb can chime in

File: src/core.js [hunk 1]
```diff
@@ -49,8 +49,13 @@ var PROGRAM_CONSTRUCTORS = {
 // Creates a single application object asynchronously
 // with a gl context, a camera, a program, a scene, and an event system.
 export async function PhiloGL(canvasId, opt = {}) {
+  // ** We can't merge DEFAULT_OPTS and opt in this way; it's not a
+  // deep merge. For example, if opt.camera doesn't define fov, then
+  // the result won't have camera.fov defined. **
+  // opt = {...DEFAULT_OPTS, ...opt};
+
   // rye: TODO- use lodash.defaultsDeep instead of $merge.
```
Comment (ibgreen) [inline src/core.js]: I see. I definitely want to get rid of the jquery emulation code, especially as a lot of it is covered by ES6 constructs.

But for simplicity of build etc, I also want to limit (eliminate) dependencies on other libraries, like lodash. 

If we do need a deep merge function, we could just rename the current function from $.merge to deepMerge, add some JSDoc to it, and put it in a utils.js file.

File: src/core.js [hunk 2]
```diff
@@ -99,18 +105,18 @@ export async function PhiloGL(canvasId, opt = {}) {
 }
 
 async function loadPrograms(app, programDescriptors) {
+  let programs = {};
   const programPromises = programDescriptors.map(async (programOpts, i) => {
     const asyncProgramConstructor = programOpts.from;
     assert(asyncProgramConstructor in PROGRAM_CONSTRUCTORS);
-    const program = await Program[asyncProgramConstructor]({
+    const program = await Program[PROGRAM_CONSTRUCTORS[asyncProgramConstructor]]({
```
Comment (ibgreen) [inline src/core.js]: This looks strange. Should it be:

File: src/scene.js [hunk 3]
```diff
@@ -101,10 +104,12 @@ export default class Scene {
   // Setup the lighting system: ambient, directional, point lights.
   setupLighting(program) {
     // Setup Lighting
-    const {
```
Comment (ibgreen) [inline src/scene.js]: Make two declarations, keep unchanged variables as `const`?

**Comment by ibgreen** [comment]
@wwwtyro This looks great. And no, it is not critical to have all examples working at this time, so I'll pull this as is.

That said, it would be interesting to know if any of the examples are dependent on the recursiveLoad functionality that we removed. If not, updating them should be quick?

**Comment by wwwtyro** [comment]
@ibgreen Thanks! 

It's been really difficult for me to estimate times on fixing these. The errors we get right now are often difficult to decode, so even small fixes can take a while to puzzle out. My best guess is 2 evenings, or roughly four hours. I'm happy to start working on them, I just didn't want to burn hours unless it's something you guys were sure you wanted. Let me know! :+1:

**Comment by ibgreen** [comment]
No worries, I have something much more fun! I just pushed a quick rewrite (non-working) of example 1 illustrating how I want application code to work after we remove the need for the PhiloGL main object and its onLoad callback. 

Take a quick look at my changes, compare them with the old code and let me know if you can see how we can make the new code work. I think we have completed almost all the "heavy lifting", we just need to put the pieces together now.

**Comment by wwwtyro** [comment]
Cool! I'll take a look right now. :grinning:

**Comment by wwwtyro** [comment]
@ibgreen This looks so, so nice. I think to root out the details I'd need to go ahead and start working through it, but I don't expect to run into any intractable issues. :+1:

**Pull request merged**